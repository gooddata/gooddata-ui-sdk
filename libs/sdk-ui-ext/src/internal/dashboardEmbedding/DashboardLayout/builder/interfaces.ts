// (C) 2019-2021 GoodData Corporation
import {
    IFluidLayoutColumnBuilder,
    IFluidLayoutRowBuilder,
    IFluidLayoutBuilder,
    FluidLayoutRowsSelector,
    FluidLayoutColumnsSelector,
    FluidLayoutModifications,
    FluidLayoutRowModifications,
    FluidLayoutColumnModifications,
} from "@gooddata/sdk-backend-spi";
import { ObjRef } from "@gooddata/sdk-model";
import {
    IDashboardViewLayout,
    IDashboardViewLayoutColumn,
    IDashboardViewLayoutRow,
} from "../interfaces/dashboardLayout";
import {
    IDashboardViewLayoutColumnFacade,
    IDashboardViewLayoutColumnsFacade,
    IDashboardViewLayoutFacade,
    IDashboardViewLayoutRowFacade,
    IDashboardViewLayoutRowsFacade,
} from "../facade/interfaces";
import { KpiWidgetBuilder, InsightWidgetBuilder } from "@gooddata/sdk-backend-base";

/**
 * Represents a query to select a subset of layout rows.
 *
 * @alpha
 */
export type DashboardViewLayoutRowsSelector<TContent> = FluidLayoutRowsSelector<
    TContent,
    IDashboardViewLayoutRow<TContent>,
    IDashboardViewLayoutRowFacade<TContent>,
    IDashboardViewLayoutRowsFacade<TContent>
>;

/**
 * Represents a query to select a subset of row columns.
 *
 * @alpha
 */
export type DashboardViewLayoutColumnsSelector<TContent> = FluidLayoutColumnsSelector<
    TContent,
    IDashboardViewLayoutColumn<TContent>,
    IDashboardViewLayoutColumnFacade<TContent>,
    IDashboardViewLayoutColumnsFacade<TContent>
>;

/**
 * Represents a callback to modify the layout.
 *
 * @alpha
 */
export type DashboardViewLayoutModifications<TContent> = FluidLayoutModifications<
    TContent,
    IDashboardViewLayout<TContent>,
    IDashboardViewLayoutRow<TContent>,
    IDashboardViewLayoutColumn<TContent>,
    IDashboardViewLayoutRowFacade<TContent>,
    IDashboardViewLayoutRowsFacade<TContent>,
    IDashboardViewLayoutColumnFacade<TContent>,
    IDashboardViewLayoutColumnsFacade<TContent>,
    IDashboardViewLayoutFacade<TContent>,
    IDashboardViewLayoutColumnBuilder<TContent>,
    IDashboardViewLayoutRowBuilder<TContent>,
    IDashboardViewLayoutBuilder<TContent>
>;

/**
 * Represents a callback to modify the layout row.
 *
 * @alpha
 */
export type DashboardViewLayoutRowModifications<TContent> = FluidLayoutRowModifications<
    TContent,
    IDashboardViewLayoutColumn<TContent>,
    IDashboardViewLayoutRow<TContent>,
    IDashboardViewLayoutRowFacade<TContent>,
    IDashboardViewLayoutColumnFacade<TContent>,
    IDashboardViewLayoutColumnsFacade<TContent>,
    IDashboardViewLayoutColumnBuilder<TContent>,
    IDashboardViewLayoutRowBuilder<TContent>
>;

/**
 * Represents a callback to modify the layout column.
 *
 * @alpha
 */
export type DashboardViewLayoutColumnModifications<TContent> = FluidLayoutColumnModifications<
    TContent,
    IDashboardViewLayoutColumn<TContent>,
    IDashboardViewLayoutColumnFacade<TContent>,
    IDashboardViewLayoutColumnBuilder<TContent>
>;
/**
 * Builder for convenient creation or transformation of any {@link IDashboardViewLayoutColumn}.
 *
 * @alpha
 */
export interface IDashboardViewLayoutColumnBuilder<TContent>
    extends IFluidLayoutColumnBuilder<
        TContent,
        IDashboardViewLayoutColumn<TContent>,
        IDashboardViewLayoutColumnFacade<TContent>
    > {
    newInsightWidget(insight: ObjRef, create?: (builder: InsightWidgetBuilder) => InsightWidgetBuilder): this;
    modifyInsightWidget(modify: (builder: InsightWidgetBuilder) => InsightWidgetBuilder): this;
    newKpiWidget(measure: ObjRef, create?: (builder: KpiWidgetBuilder) => KpiWidgetBuilder): this;
    modifyKpiWidget(modify: (builder: KpiWidgetBuilder) => KpiWidgetBuilder): this;
}

/**
 * Builder for convenient creation or transformation of any {@link IDashboardViewLayoutRow}.
 *
 * @alpha
 */
export type IDashboardViewLayoutRowBuilder<TContent> = IFluidLayoutRowBuilder<
    TContent,
    IDashboardViewLayoutRow<TContent>,
    IDashboardViewLayoutColumn<TContent>,
    IDashboardViewLayoutRowFacade<TContent>,
    IDashboardViewLayoutColumnFacade<TContent>,
    IDashboardViewLayoutColumnsFacade<TContent>,
    IDashboardViewLayoutColumnBuilder<TContent>
>;

/**
 * Builder for convenient creation or transformation of any {@link IDashboardViewLayout}.
 * The provided layout is not touched in any way, all operations performed on the layout are immutable.
 *
 * @alpha
 */
export type IDashboardViewLayoutBuilder<TContent> = IFluidLayoutBuilder<
    TContent,
    IDashboardViewLayout<TContent>,
    IDashboardViewLayoutRow<TContent>,
    IDashboardViewLayoutColumn<TContent>,
    IDashboardViewLayoutRowFacade<TContent>,
    IDashboardViewLayoutRowsFacade<TContent>,
    IDashboardViewLayoutColumnFacade<TContent>,
    IDashboardViewLayoutColumnsFacade<TContent>,
    IDashboardViewLayoutFacade<TContent>,
    IDashboardViewLayoutColumnBuilder<TContent>,
    IDashboardViewLayoutRowBuilder<TContent>
>;
