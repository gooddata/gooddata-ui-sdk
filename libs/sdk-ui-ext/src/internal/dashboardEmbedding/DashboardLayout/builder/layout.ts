// (C) 2019-2021 GoodData Corporation
import invariant from "ts-invariant";
import { isFluidLayout, FluidLayoutBuilder } from "@gooddata/sdk-backend-spi";
import {
    IDashboardViewLayoutColumnFacade,
    IDashboardViewLayoutColumnsFacade,
    IDashboardViewLayoutFacade,
    IDashboardViewLayoutRowFacade,
    IDashboardViewLayoutRowsFacade,
} from "../facade/interfaces";
import { DashboardViewLayoutFacade } from "../facade/layout";
import {
    IDashboardViewLayoutBuilder,
    IDashboardViewLayoutColumnBuilder,
    IDashboardViewLayoutRowBuilder,
} from "./interfaces";
import { DashboardViewLayoutRowBuilder } from "./row";
import {
    IDashboardViewLayout,
    IDashboardViewLayoutColumn,
    IDashboardViewLayoutRow,
} from "../interfaces/dashboardLayout";

/**
 * @alpha
 */
export class DashboardViewLayoutBuilder<TContent>
    extends FluidLayoutBuilder<
        TContent,
        IDashboardViewLayout<TContent>,
        IDashboardViewLayoutRow<TContent>,
        IDashboardViewLayoutColumn<TContent>,
        IDashboardViewLayoutRowFacade<TContent>,
        IDashboardViewLayoutRowsFacade<TContent>,
        IDashboardViewLayoutColumnFacade<TContent>,
        IDashboardViewLayoutColumnsFacade<TContent>,
        IDashboardViewLayoutFacade<TContent>,
        IDashboardViewLayoutColumnBuilder<TContent>,
        IDashboardViewLayoutRowBuilder<TContent>
    >
    implements IDashboardViewLayoutBuilder<TContent> {
    protected constructor(
        protected layoutFacade: IDashboardViewLayoutFacade<TContent>,
        protected layoutFacadeConstructor: (
            layout: IDashboardViewLayout<TContent>,
        ) => IDashboardViewLayoutFacade<TContent>,
        protected getRowBuilder: (rowIndex: number) => IDashboardViewLayoutRowBuilder<TContent>,
    ) {
        super(layoutFacade, layoutFacadeConstructor, getRowBuilder);
    }

    /**
     * Creates an instance of DashboardViewLayoutBuilder for particular layout.
     *
     * @param layout - layout to modify
     */
    public static for<TContent>(
        layout: IDashboardViewLayout<TContent>,
    ): DashboardViewLayoutBuilder<TContent> {
        invariant(isFluidLayout(layout), "Provided data must be IDashboardViewLayout!");
        const layoutFacade = DashboardViewLayoutFacade.for(layout);

        const layoutBuilder = new DashboardViewLayoutBuilder(
            layoutFacade,
            DashboardViewLayoutFacade.for,
            getRowBuilder,
        );

        function getRowBuilder(rowIndex: number): IDashboardViewLayoutRowBuilder<TContent> {
            return DashboardViewLayoutRowBuilder.for(layoutBuilder, rowIndex);
        }

        return layoutBuilder;
    }

    /**
     * Creates an instance of DashboardViewLayoutBuilder with empty layout.
     */
    public static forNewLayout<TContent>(): DashboardViewLayoutBuilder<TContent> {
        const emptyFluidLayout: IDashboardViewLayout<TContent> = {
            type: "fluidLayout",
            rows: [],
        };
        return DashboardViewLayoutBuilder.for(emptyFluidLayout);
    }
}
