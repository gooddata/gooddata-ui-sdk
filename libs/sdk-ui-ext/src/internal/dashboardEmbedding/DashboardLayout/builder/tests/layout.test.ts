// (C) 2019-2021 GoodData Corporation
import { DashboardViewLayoutBuilder } from "../layout";
import { IDashboardViewLayout } from "../../interfaces/dashboardLayout";
import { createEmptyFluidLayoutBuilder, defaultColumnXlSize } from "./utils";
import { idRef } from "@gooddata/sdk-model";
import { newInsightWidget, newKpiWidget } from "@gooddata/sdk-backend-base";

describe("DashboardViewLayoutBuilder", () => {
    describe("constructors", () => {
        describe(".for()", () => {
            it("should use the provided layout", () => {
                const layout: IDashboardViewLayout<any> = {
                    type: "fluidLayout",
                    rows: [{ columns: [] }],
                };

                expect(DashboardViewLayoutBuilder.for(layout).build()).toBe(layout);
            });

            it("should throw error, when the layout is damaged", () => {
                const boom = () => {
                    const layoud: any = {
                        type: "yolo",
                        rowz: [{ columnz: [] }],
                    };

                    DashboardViewLayoutBuilder.for(layoud);
                };

                expect(boom).toThrowErrorMatchingSnapshot();
            });
        });
    });

    it("should resize columns with particular insight", () => {
        const tableInsightRef = idRef("tableInsight");
        const barInsightRef = idRef("barInsight");
        const columnInsightRef = idRef("columnInsight");

        const contentRows = [
            [
                newInsightWidget(tableInsightRef, (w) => w.title("Table widget").ref(idRef("widget1"))),
                newInsightWidget(barInsightRef, (w) => w.title("Bar widget").ref(idRef("widget2"))),
                newInsightWidget(columnInsightRef, (w) => w.title("Column widget").ref(idRef("widget3"))),
            ],
            [newInsightWidget(tableInsightRef, (w) => w.title("Table widget 2").ref(idRef("widget4")))],
        ];

        const layoutBuilder = createEmptyFluidLayoutBuilder();

        for (const contentRow of contentRows) {
            layoutBuilder.addRow((row) => {
                for (const content of contentRow) {
                    row.addColumn(defaultColumnXlSize, (col) => col.content(content));
                }
                return row;
            });
        }

        const layoutBefore = layoutBuilder.build();
        const layoutAfter = layoutBuilder
            .modifyRows((r) =>
                r.modifyColumns(
                    (column) => column.size({ xl: { widthAsGridColumnsCount: 2, heightAsRatio: 25 } }),
                    (columns) => columns.filter((column) => column.hasWidgetWithInsightRef(tableInsightRef)),
                ),
            )
            .build();

        expect(layoutBefore).toMatchSnapshot("before");
        expect(layoutAfter).toMatchSnapshot("after");
    });

    it("should remove all columns with kpi widget", () => {
        const insightRef = idRef("insightRef");
        const measureRef = idRef("measure");
        const insightWidgetRef = idRef("insightWidget");
        const kpiWidgetRef = idRef("kpiWidget");

        const contentRows = [
            [
                newInsightWidget(insightRef, (w) => w.title("Insight widget").ref(insightWidgetRef)),
                newKpiWidget(measureRef, (w) => w.title("Kpi widget").ref(kpiWidgetRef)),
            ],
            [newKpiWidget(measureRef, (w) => w.title("Kpi widget").ref(kpiWidgetRef))],
            [newInsightWidget(insightRef, (w) => w.title("Insight widget").ref(insightWidgetRef))],
        ];

        const layoutBuilder = createEmptyFluidLayoutBuilder();

        for (const contentRow of contentRows) {
            layoutBuilder.addRow((row) => {
                for (const content of contentRow) {
                    row.addColumn(defaultColumnXlSize, (col) => col.content(content));
                }
                return row;
            });
        }

        const layoutBefore = layoutBuilder.build();
        const layoutAfter = layoutBuilder
            .modifyRows((r) =>
                r.removeColumns((columns) => {
                    return columns.filter((column) => column.hasKpiWidgetContent());
                }),
            )
            .removeEmptyRows()
            .build();

        expect(layoutBefore).toMatchSnapshot("before");
        expect(layoutAfter).toMatchSnapshot("after");
    });
});
