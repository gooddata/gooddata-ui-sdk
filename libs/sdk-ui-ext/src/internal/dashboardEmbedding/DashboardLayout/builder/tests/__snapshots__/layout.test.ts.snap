// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardLayoutBuilder .addSection() should add a section to the end of the layout by default: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Added section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .addSection() should add a section to the end of the layout by default: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .addSection() should add a section to the provided index: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Added section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .addSection() should add a section to the provided index: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySection() should modify the section: after 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Modified section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySection() should modify the section: before 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Original section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySection() should throw error, when the section at provided index does not exist 1`] = `"Cannot modify the section - section at index 0 does not exist!"`;

exports[`DashboardLayoutBuilder .modifySections() should modify all sections by default: after 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Modified section 0",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Modified section 1",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should modify all sections by default: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should modify selected section only: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Modified section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should modify selected section only: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Original section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should modify selected sections only: after 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Modified section 0",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Modified section 2",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should modify selected sections only: before 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Original section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Original section",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should not modify any section, when selection result is undefined: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .modifySections() should not modify any section, when selection result is undefined: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .moveSection() should move the section: after 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Row 1",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row 0",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row 2",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .moveSection() should move the section: before 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Row 0",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row 1",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row 2",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .moveSection() should throw error, when the section at provided index does not exist 1`] = `"Cannot move the section - section at index 0 does not exist!"`;

exports[`DashboardLayoutBuilder .removeSection() should remove the section from the layout: after 1`] = `
Object {
  "sections": Array [],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSection() should remove the section from the layout: before 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Row to remove",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSection() should throw error, when the section at provided index does not exist 1`] = `"Cannot remove the section - section at index 0 does not exist!"`;

exports[`DashboardLayoutBuilder .removeSections() should not remove any section, when selection result is undefined: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should not remove any section, when selection result is undefined: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should remove all sections by default: after 1`] = `
Object {
  "sections": Array [],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should remove all sections by default: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should remove selected section only: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should remove selected section only: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row to remove",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should remove selected sections only: after 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Row to keep 1",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row to keep 2",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .removeSections() should remove selected sections only: before 1`] = `
Object {
  "sections": Array [
    Object {
      "header": Object {
        "title": "Row to keep 1",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "header": Object {
        "title": "Row to keep 2",
      },
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .size() should update size by callback 1`] = `
Object {
  "sections": Array [],
  "size": Object {
    "gridWidth": 12,
    "heightAsRatio": 50,
  },
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .size() should update size by callback returning undefined 1`] = `
Object {
  "sections": Array [],
  "size": undefined,
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .size() should update size by undefined 1`] = `
Object {
  "sections": Array [],
  "size": undefined,
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder .size() should update size by value 1`] = `
Object {
  "sections": Array [],
  "size": Object {
    "gridWidth": 12,
    "heightAsRatio": 50,
  },
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder constructors .for() should throw error, when the layout is damaged 1`] = `"Provided data must be IDashboardLayout!"`;

exports[`DashboardLayoutBuilder constructors .forNewLayout() should create new empty layout 1`] = `
Object {
  "sections": Array [],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder should remove all items with kpi widget: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "insightRef",
            },
            "ref": Object {
              "identifier": "insightWidget",
            },
            "title": "Insight widget",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "insightRef",
            },
            "ref": Object {
              "identifier": "insightWidget",
            },
            "title": "Insight widget",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder should remove all items with kpi widget: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "insightRef",
            },
            "ref": Object {
              "identifier": "insightWidget",
            },
            "title": "Insight widget",
            "type": "insight",
          },
        },
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "kpi": Object {
              "comparisonType": "none",
              "metric": Object {
                "identifier": "measure",
              },
            },
            "ref": Object {
              "identifier": "kpiWidget",
            },
            "title": "Kpi widget",
            "type": "kpi",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "kpi": Object {
              "comparisonType": "none",
              "metric": Object {
                "identifier": "measure",
              },
            },
            "ref": Object {
              "identifier": "kpiWidget",
            },
            "title": "Kpi widget",
            "type": "kpi",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "insightRef",
            },
            "ref": Object {
              "identifier": "insightWidget",
            },
            "title": "Insight widget",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder should resize items with particular insight: after 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 2,
              "heightAsRatio": 25,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "tableInsight",
            },
            "ref": Object {
              "identifier": "widget1",
            },
            "title": "Table widget",
            "type": "insight",
          },
        },
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "barInsight",
            },
            "ref": Object {
              "identifier": "widget2",
            },
            "title": "Bar widget",
            "type": "insight",
          },
        },
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "columnInsight",
            },
            "ref": Object {
              "identifier": "widget3",
            },
            "title": "Column widget",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 2,
              "heightAsRatio": 25,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "tableInsight",
            },
            "ref": Object {
              "identifier": "widget4",
            },
            "title": "Table widget 2",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;

exports[`DashboardLayoutBuilder should resize items with particular insight: before 1`] = `
Object {
  "sections": Array [
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "tableInsight",
            },
            "ref": Object {
              "identifier": "widget1",
            },
            "title": "Table widget",
            "type": "insight",
          },
        },
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "barInsight",
            },
            "ref": Object {
              "identifier": "widget2",
            },
            "title": "Bar widget",
            "type": "insight",
          },
        },
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "columnInsight",
            },
            "ref": Object {
              "identifier": "widget3",
            },
            "title": "Column widget",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
    Object {
      "items": Array [
        Object {
          "size": Object {
            "xl": Object {
              "gridWidth": 12,
              "heightAsRatio": 50,
            },
          },
          "type": "IDashboardLayoutItem",
          "widget": Object {
            "description": "",
            "drills": Array [],
            "ignoreDashboardFilters": Array [],
            "insight": Object {
              "identifier": "tableInsight",
            },
            "ref": Object {
              "identifier": "widget4",
            },
            "title": "Table widget 2",
            "type": "insight",
          },
        },
      ],
      "type": "IDashboardLayoutSection",
    },
  ],
  "type": "IDashboardLayout",
}
`;
